src/gsym/parser.rs

```diff
@@ -0,0 +1,323 @@
+        self.header.num_addrs as usize
+    }
+
+    /// Get the address of the an entry in the Address Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
+    #[inline(always)]
+    pub fn addr_at(&self, idx: usize) -> u64 {
+        assert!(idx < self.header.num_addrs as usize, "invalid index");
+        let off = idx * self.header.addr_off_size as usize;
+        let mut addr = 0u64;
+        let mut shift = 0;
+        for d in &self.addr_tab[off..(off + self.header.addr_off_size as usize)] {
+            addr |= (*d as u64) << shift;
+            shift += 8;
+        }
+        addr += self.header.base_address;
+        addr
+    }
+
+    /// Get the AddressInfo of an address given by an index.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
+    #[inline(always)]
+    pub fn addr_info(&self, idx: usize) -> AddressInfo {
+        assert!(idx < self.header.num_addrs as usize, "invalid index");
+        let off = idx * 4;
+        let ad_off = decode_uword(&self.addr_data_off_tab[off..]) as usize;
+        let size = decode_uword(&self.raw_data[ad_off..]);
+        let name = decode_uword(&self.raw_data[ad_off + 4..]);
+        AddressInfo {
+            size,
+            name,
+            data: &self.raw_data[ad_off + 8..],
+        }
+    }
+
+    /// Get the string at the given offset from the String Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid offset.
+    #[inline(always)]
+    pub fn get_str(&self, off: usize) -> &str {
+        assert!(self.str_tab[self.str_tab.len() - 1] == 0);
+        assert!(off < self.str_tab.len());
+        unsafe {
+            CStr::from_ptr(self.str_tab[off..].as_ptr() as *const i8)
+                .to_str()
+                .unwrap()
+        }
+    }
+}
+
+/// Find the index of an entry in the address table most likely
+/// containing the given address.
+///
+/// The callers should check the respective `AddressInfo` to make sure
+/// it is what they request for.
+pub fn find_address(ctx: &GsymContext, addr: u64) -> usize {
+    let mut left = 0;
+    let mut right = ctx.num_addresses();
+
+    if right == 0 {
+        return 0;
+    }
+    if addr < ctx.addr_at(0) {
+        return 0;
+    }
+
+    while (left + 1) < right {
+        let v = (left + right) / 2;
+        let cur_addr = ctx.addr_at(v);
+
+        if addr == cur_addr {
+            return v;
+        }
+        if addr < cur_addr {
+            right = v;
+        } else {
+            left = v;
+        }
+    }
+    left
+}
+
+/// Parse AddressData.
+///
+/// AddressDatas are items following AndressInfo.
+/// [`GsymContext::addr_info()`] returns the raw data of AddressDatas as a
+/// slice at [`AddressInfo::data`].
+///
+/// # Arguments
+///
+/// * `data` - is the slice from AddressInfo::data.
+///
+/// Returns a vector of [`AddressData`].
+pub fn parse_address_data(data: &[u8]) -> Vec<AddressData> {
```

Comment:
This function will be used in the next patch.

So far, this PR deals with the parser.  It doesn't expose all data to the API yet.  Next PR will work on line numbers.  It will use this function.